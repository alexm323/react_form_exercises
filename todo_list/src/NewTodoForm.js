import React, { useState } from 'react';

const NewTodoForm = ({ addTodo }) => {
    const INITIAL_STATE = {
        task: ''
    }
    const [formData, setFormData] = useState(INITIAL_STATE)
    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        addTodo({ ...formData });
        setFormData({ todo: "" });

    }
    return (
        <form onSubmit={handleSubmit}>
            <label forHtml='todo'>todo</label>
            <input id='todo' aria-label="todo" name='todo' type="text" value={formData.todo} onChange={handleChange} />
            <button>Add Todo</button>
        </form>
    )
}

export default NewTodoForm;