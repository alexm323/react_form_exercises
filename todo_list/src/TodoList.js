import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';

const TodoList = () => {
    const [todos, setTodos] = useState([])

    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, { ...newTodo, id: uuid() }])
    }
    const removeTodo = (id) => {
        setTodos(todos => todos.filter(todo => todo.id !== id))
    }
    return (
        <div>
            <NewTodoForm addTodo={addTodo} />
            <h2>Todo list:</h2>
            <ul>
                {todos.map(todo => <Todo task={todo.todo} handleRemove={removeTodo} id={todo.id} key={todo.id} />)}
            </ul>

        </div>

    )
}

export default TodoList;

// TodoList - this component should render the NewTodoForm component and should render the list of Todo components. Place your state that contains all of the todos in this component.