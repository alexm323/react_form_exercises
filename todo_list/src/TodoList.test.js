import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';
import NewTodoForm from './NewTodoForm';

// smokebox test
it('renders without crashing', () => {
    render(<TodoList />);
})
// snapshot
it('matches snapshot', () => {
    const { asFragment } = render(<NewTodoForm />);
    expect(asFragment()).toMatchSnapshot();

})
// add and remove a todo
it('adds a todo to the DOM', () => {
    const todoList = render(<TodoList />);
    const taskInput = todoList.getByLabelText("todo");
    const submitBtn = todoList.getByText("Add Todo");

    expect(todoList.queryByText("Remove Task")).not.toBeInTheDocument()

    fireEvent.change(taskInput, { target: { value: "Buy potatoes" } })
    fireEvent.click(submitBtn)

    expect(todoList.queryByText("Remove Task")).toBeInTheDocument();
    expect(todoList.queryByText("Buy potatoes")).toBeInTheDocument();

    const taskRemoveBtn = todoList.getByText("Remove Task");
    fireEvent.click(taskRemoveBtn);
    expect(todoList.queryByText("Buy potatoes")).not.toBeInTheDocument();

})