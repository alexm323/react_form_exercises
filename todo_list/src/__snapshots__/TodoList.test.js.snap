// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      forhtml="todo"
    >
      todo
    </label>
    <input
      aria-label="todo"
      id="todo"
      name="todo"
      type="text"
      value=""
    />
    <button>
      Add Todo
    </button>
  </form>
</DocumentFragment>
`;
